### DISCUSSION TITLE
PPOCRLabel sometimes returns crop_img upsidedown

### DISCUSSION THREAD

**Discussion by l03126** [PaddlePaddle/PaddleOCR#14377] [Q&A]
I'm using PPOCRLabel to annotate images for rec training. Most of the cropped images are fine when exported. But some of it show upside down even though it was straight when annotating. 

To my knowledge, this function may be trying to deskew an image automatically.

```
def get_rotate_crop_image(img, points):
    # Use Green's theory to judge clockwise or counterclockwise
    # author: biyanhua
    d = 0.0
    for index in range(-1, 3):
        d += (
            -0.5
            * (points[index + 1][1] + points[index][1])
            * (points[index + 1][0] - points[index][0])
        )
    if d < 0:  # counterclockwise
        tmp = np.array(points)
        points[1], points[3] = tmp[3], tmp[1]

    try:
        img_crop_width = int(
            max(
                np.linalg.norm(points[0] - points[1]),
                np.linalg.norm(points[2] - points[3]),
            )
        )
        img_crop_height = int(
            max(
                np.linalg.norm(points[0] - points[3]),
                np.linalg.norm(points[1] - points[2]),
            )
        )
        pts_std = np.float32(
            [
                [0, 0],
                [img_crop_width, 0],
                [img_crop_width, img_crop_height],
                [0, img_crop_height],
            ]
        )
        M = cv2.getPerspectiveTransform(points, pts_std)
        dst_img = cv2.warpPerspective(
            img,
            M,
            (img_crop_width, img_crop_height),
            borderMode=cv2.BORDER_REPLICATE,
            flags=cv2.INTER_CUBIC,
        )
        dst_img_height, dst_img_width = dst_img.shape[0:2]
        if dst_img_height * 1.0 / dst_img_width >= 1.5:
            dst_img = np.rot90(dst_img)
        return dst_img
    except Exception as e:
        print(e)
```

 Is there a way to specify to not try and crop the image similar to how we specify the language?
` python PPOCRLabel.py --lang en
`
Or is the only way to do it is by manually editing the code?

**Comment by GreatV** [comment]
The issue of some cropped images being upside down in PPOCRLabel is likely due to the automatic rotation logic in the `get_rotate_crop_image` function. Specifically, the condition:

```python
if dst_img_height * 1.0 / dst_img_width >= 1.5:
    dst_img = np.rot90(dst_img)
```

This logic rotates the cropped image by 90 degrees if the height-to-width ratio of the cropped image exceeds 1.5. This behavior is intended to ensure the text is oriented correctly for recognition, but in your use case, it seems to be causing undesired rotations.

### Solutions:

#### 1. **Manually Edit the Code to Disable Rotation**
The simplest solution is to disable the automatic rotation logic by modifying the code. Comment out or remove the following block in the `get_rotate_crop_image` function:

```python
if dst_img_height * 1.0 / dst_img_width >= 1.5:
    dst_img = np.rot90(dst_img)
```

This will prevent any automatic rotations and ensure the cropped images maintain their original orientation.

#### 2. **Add a Command-Line Parameter to Control Rotation**
If you want more flexibility (e.g., to toggle rotation on or off without modifying the code each time), you can introduce a command-line parameter to control this behavior. Here's how:

1. Update the `PPOCRLabel.py` script to accept a new parameter, for example, `--disable_rotation`.
2. Pass this parameter to the `get_rotate_crop_image` function.
3. Modify the function to check this parameter before applying the rotation.

Example modification:

```python
def get_rotate_crop_image(img, points, disable_rotation=False):
    # Use Green's theory to judge clockwise or counterclockwise
    d = 0.0
    for index in range(-1, 3):
        d += (
            -0.5
            * (points[index + 1][1] + points[index][1])
            * (points[index + 1][0] - points[index][0])
        )
    if d < 0:  # counterclockwise
        tmp = np.array(points)
        points[1], points[3] = tmp[3], tmp[1]

    try:
        img_crop_width = int(
            max(
                np.linalg.norm(points[0] - points[1]),
                np.linalg.norm(points[2] - points[3]),
            )
        )
        img_crop_height = int(
            max(
                np.linalg.norm(points[0] - points[3]),
                np.linalg.norm(points[1] - points[2]),
            )
        )
        pts_std = np.float32(
            [
                [0, 0],
                [img_crop_width, 0],
                [img_crop_width, img_crop_height],
                [0, img_crop_height],
            ]
        )
        M = cv2.getPerspectiveTransform(points, pts_std)
        dst_img = cv2.warpPerspective(
            img,
            M,
            (img_crop_width, img_crop_height),
            borderMode=cv2.BORDER_REPLICATE,
            flags=cv2.INTER_CUBIC,
        )
        dst_img_height, dst_img_width = dst_img.shape[0:2]
        if not disable_rotation and dst_img_height * 1.0 / dst_img_width >= 1.5:
            dst_img = np.rot90(dst_img)
        return dst_img
    except Exception as e:
        print(e)
```

Then, call this function with `disable_rotation=True` if the parameter is specified.

#### 3. **File-Specific Rotation Control**
If only specific images exhibit this issue, you could handle it at the annotation level. For example, store metadata about each image's orientation during annotation and use it to decide whether to apply rotation.

#### 4. **Submit an Issue or Feature Request**
If you believe this is a common problem or enhancement, consider submitting an issue or feature request to the [PaddleOCR GitHub repository](https://github.com/PaddlePaddle/PaddleOCR/issues). You could suggest adding a built-in option to enable or disable rotation during the cropping process.

---

### Additional Notes:
- If you're using the `--lang` parameter for specifying the language (e.g., `python PPOCRLabel.py --lang en`), there is currently no built-in parameter to control rotation. The solutions above involve extending the script functionality.
- Ensure you test the modified function with different datasets to confirm the desired behavior.

If you need further clarification or assistance with implementing these changes, let me know!

---
*Response generated by [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*
